src/random.rs

```diff
@@ -1,83 +1,51 @@
-use rand::seq::SliceRandom;
+use rand::{prelude::ThreadRng, seq::SliceRandom};
+use std::{borrow::Cow, iter};
 
 const RECURSION_LIMIT: u8 = 1;
 
-trait Type: Send + Sync {
-    fn num_generic_params(&self) -> usize;
-    fn stringify(&self, generic_params: &[String]) -> String;
-}
-
-impl Type for &'static str {
-    fn num_generic_params(&self) -> usize {
-        0
-    }
-
-    fn stringify(&self, _generic_params: &[String]) -> String {
-        (*self).to_owned()
-    }
-}
-
-impl Type for (&'static str, &'static str) {
-    fn num_generic_params(&self) -> usize {
-        1
-    }
-
-    fn stringify(&self, generic_params: &[String]) -> String {
-        format!("{}{}{}", self.0, generic_params[0], self.1)
-    }
-}
-
-impl Type for (&'static str, &'static str, &'static str) {
-    fn num_generic_params(&self) -> usize {
-        2
-    }
-
-    fn stringify(&self, generic_params: &[String]) -> String {
-        format!("{}{}{}{}{}", self.0, generic_params[0], self.1, generic_params[1], self.2)
-    }
-}
-
-static TYPES: &[&dyn Type] = &[
-    &"_",
-    &"bool",
-    &"char",
-    &"i8",
-    &"i16",
-    &"i32",
-    &"i64",
-    &"isize",
-    &"u8",
-    &"u16",
-    &"u32",
-    &"u64",
-    &"usize",
-    &"f32",
-    &"f64",
-    &"&str",
-    &"String",
-    &"()",
-    &("&", ""),
-    &("&mut ", ""),
-    &("[", "]"),
-    &("Box<", ">"),
-    &("Vec<", ">"),
-    &("HashSet<", ">"),
-    &("Result<", ", ", ">"),
-    &("HashMap<", ", ", ">"),
+const TYPES: &[&[&str]] = &[
+    &["_"],
+    &["bool"],
+    &["char"],
+    &["i8"],
+    &["i16"],
+    &["i32"],
+    &["i64"],
+    &["isize"],
+    &["u8"],
+    &["u16"],
+    &["u32"],
+    &["u64"],
+    &["usize"],
+    &["f32"],
+    &["f64"],
+    &["&str"],
+    &["String"],
+    &["()"],
+    &["&", ""],
+    &["&mut ", ""],
+    &["[", "]"],
+    &["Box<", ">"],
+    &["Vec<", ">"],
+    &["HashSet<", ">"],
+    &["Result<", ", ", ">"],
+    &["HashMap<", ", ", ">"],
 ];
 
 pub fn random_type() -> String {
-    random_type_depth(0)
+    random_type_depth(0, &mut rand::thread_rng())
 }
 
-fn random_type_depth(depth: u8) -> String {
-    let ty = TYPES.choose(&mut rand::thread_rng()).unwrap();
-
-    if depth == RECURSION_LIMIT {
-        ty.stringify(&(0..ty.num_generic_params()).map(|_| "_".to_owned()).collect::<Vec<_>>())
-    } else {
-        ty.stringify(
-            &(0..ty.num_generic_params()).map(|_| random_type_depth(depth + 1)).collect::<Vec<_>>(),
-        )
-    }
+fn random_type_depth(depth: u8, rng: &mut ThreadRng) -> String {
```

Comment:
However, I find the new implementation of this function to be rather hard to read (both before & after force-push). Making the rng a parameter is nice, but the iterator magic below is far from obvious. I've partially made sense of it after reading through it a few times, but could you try making it a bit more legible?

If that's too vague I'll try to come up with an alternative formulation myself.

Reply:
Yeah, I suppose this is unnecessarily complicated.

Actually, apparently Rust 1.56 is going to stabilize `intersperse_with` which would allow writing this thus:

```rust
fn random_type_depth(depth: u8, rng: &mut ThreadRng) -> String {
    let &ty = TYPES.choose(rng).unwrap();
    ty.iter()
        .map(|&x| Cow::from(x))
        .intersperse_with(|| {
            if depth == RECURSION_LIMIT {
                "_".into()
            } else {
                random_type_depth(depth + 1, rng).into()
            }
        })
        .collect()
}
```

Maybe it would make sense to wait for stabilization then.

Reply:
I love that, except I'd write `Cow::Borrowed` instead of `Cow::from` (unless that's not the same thing?). Maybe even have `random_type_depth` return `arrayvec::ArrayVec<&str; 10>` and build the string after the whole recursion (in `random_type`)? Would have to make sure to update that `10` when types with more generic parameters are introduced or recursion depth is increased.

… which makes me think, there's actually more generic params on `HashMap` and `Vec` and so on now, but adding them just like that would result in the random functionality returning invalid types. Specifying invalid types, even strings that wouldn't parse as types, being possible is kind of by design but I feel the random routes shouldn't give you invalid types. This is slightly violated now via `f32` and `f64` being in the list.

So much for my brain dump; thoughts? 😄